[clang-tidy] Fix smart pointers handling in bugprone-use-after-move

PiotrZSL opened the pull request:
Removed custom handling of smart pointers and added option IgnoreNonDerefSmartPtrs to restore
previous behavior if needed.

Closes #90174

martinboehme commented:
More extensive comments on #90174 (let's continue the discussion there), but briefly:

> * Removed custom smart pointers handling (were hiding issues)

The standard smart pointers have a well-defined moved-from state (they are defined to be null after a move). Programmers may intentionally rely on this defined behavior. Warning on a (non-dereferencing) use-after-move for smart pointers as we do for other types would therefore result in false positives.

> * Changed 'move occurred here' note location to always point to 'std::move'

As discussed in more detail on #90174, the move doesn't actually occur in the `std::move()` (which is just a cast), and this can be important when sequencing rules come into play.

PiotrZSL committed:
[clang-tidy] Fix smart pointers handling in bugprone-use-after-move

Removed custom handling of smart pointers and
added option IgnoreNonDerefSmartPtrs to restore
previous behavior if needed.

PiotrZSL commented:
Rebase, bump

martinboehme reviewed (changes requested):
Sorry - I only noticed now after your rebase that this had been sitting in my review queue  for quite a while.

*  I would suggest rewording the option name. The "ignore" (`IgnoreNonDerefSmartPtrs`) reads like a negation, which makes it harder to reason about what the option does. How about `AllowMovedSmartPtrUse`? (I think maybe the "non-deref" can be omitted from the option name -- I think it's sufficient that it's explained in the documentation).

*  In https://github.com/llvm/llvm-project/issues/90174, I suggested that if we wanted to change the default behavior, we should survey users to see what their opinions / preferences are. Have you done this? If not, I think the default behavior (when the option is not set) should remain unchanged.

PiotrZSL committed:
Rename IgnoreNonDerefSmartPtrs into AllowMovedSmartPtrUse

PiotrZSL commented:
@martinboehme 
_"I suggested that if we wanted to change the default behavior, we should survey users to see what their opinions / preferences are. Have you done this? If not, I think the default behavior (when the option is not set) should remain unchanged."_

Nope, there is no reliable way to "survey users". Default behavior of checks changes every release. Some checks are getting limited, by excluding some use cases, some checks are getting extended. In this case situation is easy. Users will update Clang-tidy, and if check will report "new issues", they have an option, to fix them or look into release notes to find out what changed in check and then change configuration to revert back to previous behavior. This is how things were handled in past for other checks.

martinboehme commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/use-after-move.rst:
```diff
@@ -250,3 +251,14 @@ For example, if an additional member variable is added to ``S``, it is easy to
+   If this option is set to `true`, the check will not warn about uses of
+   ``std::unique_ptr``, ``std::shared_ptr`` that are not dereferences. This
+   can be useful if you are using these smart pointers in a way that is not
+   idiomatic, but that you know is safe. Default is `false`.
```
"This can be useful if you are using these smart pointers in a way that is not idiomatic, but that you know is safe."

I think "not idiomatic" is too strong -- I would say this is a question of style. Maybe just leave out this entire sentence (i.e. describe only the behavior)?

martinboehme commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/use-after-move-smart-ptr.cpp:
```diff
@@ -0,0 +1,59 @@
+
+void natively(std::unique_ptr<A> x) {
+  std::unique_ptr<A> tmp = std::move(x);
+  std::unique_ptr<SinkA> y2{new SinkA(std::move(x))};
```
Is there a reason (here and possibly also below) that `SinkA` needs to be referenced via a `unique_ptr` instead of simply making it a local variable of `SinkA`?

I.e. why not simply do

```cxx
SinkA sink(std::move(x));
```

since the behavior we're interested in is what happens when we move `x` into the argument for the `SinkA` constructor, and not what happens with the `SinkA` itself.

(Also nit: I'm not clear on the reason for the `y2` variable name -- a) there is no `y1`, and b) maybe just `sink` to emphasize the type?)

martinboehme commented:
> Nope, there is no reliable way to "survey users". Default behavior of checks changes every release. Some checks are getting limited, by excluding some use cases, some checks are getting extended. In this case situation is easy. Users will update Clang-tidy, and if check will report "new issues", they have an option, to fix them or look into release notes to find out what changed in check and then change configuration to revert back to previous behavior. This is how things were handled in past for other checks.

OK, in that case this sounds fine.

5chmidti reviewed (approved):
LGTM minus the other comments
